### PR TITLE
feat(broker): get workflow outcome

### LINKED ISSUE #2896: Feature request: Awaitable Workflow outcomes
**Is your feature request related to a problem? Please describe.**
At the moment it is not possible to start a workflow and be notified of the outcome.

**Describe the solution you'd like**
 In the client libs, some way to await the outcome of a workflow.

**Describe alternatives you've considered**
https://zeebe.io/blog/2019/08/zeebe-rest-affinity/

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

* Clients can create a workflow instance via a new grpc request `CreateWorkflowInstanceWithResult` and get the response when the workflow is completed.

## Related issues

closes #2896

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

File: docs/src/reference/grpc.md [hunk 1]
```diff
@@ -249,6 +250,43 @@ message CreateWorkflowInstanceResponse {
 }
 ```
 
+### CreateWorkflowInstanceWithResult RPC
+
+Similar to `CreateWorkflowInstance RPC` , ceates and starts an instance of the specified workflow.
```
Comment (Zelldon) [inline docs/src/reference/grpc.md]:
```suggestion
Similar to `CreateWorkflowInstance RPC` , creates and starts an instance of the specified workflow. 
```

File: gateway-protocol/src/main/proto/gateway.proto [hunk 2]
```diff
@@ -113,6 +113,29 @@ message CreateWorkflowInstanceResponse {
   int64 workflowInstanceKey = 4;
 }
 
+message CreateWorkflowInstanceWithResultRequest {
+  CreateWorkflowInstanceRequest request = 1;
+  // timeout in milliseconds. the request will be closed if the workflow is not completed
+  // before the requestTimeout
+  int64 requestTimeout = 2;
```
Comment (Zelldon) [inline gateway-protocol/src/main/proto/gateway.proto]: what is the default here?

File: gateway/src/main/java/io/zeebe/gateway/EndpointManager.java [hunk 3]
```diff
@@ -141,6 +144,26 @@ public void createWorkflowInstance(
         responseObserver);
   }
 
+  @Override
+  public void createWorkflowInstanceWithResult(
+      CreateWorkflowInstanceWithResultRequest request,
+      StreamObserver<CreateWorkflowInstanceWithResultResponse> responseObserver) {
+    if (request.getRequestTimeout() > 0) {
```
Comment (Zelldon) [inline gateway/src/main/java/io/zeebe/gateway/EndpointManager.java]: What is the grpc default timeout?

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Looks good one question about the timeout is this value optional? If so what will be the default?

File: gateway/src/main/java/io/zeebe/gateway/EndpointManager.java (hunk 3 above)
Comment (deepthidevaki) [reply gateway/src/main/java/io/zeebe/gateway/EndpointManager.java]: Thanks @Zelldon for fast review.
The default value for the field 'requestTimeout' in the grpc message is 0. If the value is 0, we use the same timeout that we use for other requests - default 15s, but configurable. I will make it explicit in the docs.

**Comment by deepthidevaki** [comment]
@Zelldon I have updated the docs. I didn't specify what is the default timeout, because it is configurable in the gateway.

**Comment by Zelldon** [comment]
Should we add this as well *" configurable in the Gateway" ?

**Comment by deepthidevaki** [comment]
> Should we add this as well *" configurable in the Gateway" ?

Not sure. There is no configuration specific to timeout for this request type. The timeout is configured in general for all requests.

**Review by ChrisKujawa — APPROVED** [review]
Thanks @deepthidevaki 

I would assume that it would be more understandable as you wrote "uses per default the generic request timeout which is configured at the gateway (default 15 s)"

**Comment by jwulf** [comment]
If the broker is under heavy load and the workflow does not complete in time, what happens then?

**Pull request merged**